zcash_history/src/tree.rs

```diff
@@ -640,100 +639,88 @@ mod tests {
-                }
-                for _ in 0..number {
-                    tree.truncate_leaf().expect("Failed to truncate");
-                }
-
-                TestResult::from_bool(matches!(tree.root(), EntryLink::Stored(2)))
+    proptest! {
+        #[test]
+        fn prop_there_and_back(number in 0u32..=1024) {
+            let mut tree = initial();
+            for i in 0..number {
+                tree.append_leaf(leaf(i+3)).expect("Failed to append");
+            }
+            for _ in 0..number {
+                tree.truncate_leaf().expect("Failed to truncate");
             }
-        }
 
-        fn leaf_count(number: u32) -> TestResult {
-            if !(3..=1024 * 1024).contains(&number) {
-                TestResult::discard()
-            } else {
-                let mut tree = initial();
-                for i in 1..(number-1) {
-                    tree.append_leaf(leaf(i+2)).expect("Failed to append");
-                }
+            assert_matches!(tree.root(), EntryLink::Stored(2));
+        }
 
-                TestResult::from_bool(
-                    tree.root_node().expect("no root").node.leaf_count() == number as u64
-                )
+        #[test]
+        fn prop_leaf_count(number in 3u32..=1024) {
+            let mut tree = initial();
+            for i in 1..(number-1) {
+                tree.append_leaf(leaf(i+2)).expect("Failed to append");
             }
+
+            assert_eq!(tree.root_node().expect("no root").node.leaf_count(), number as u64);
         }
 
-        fn parity(number: u32) -> TestResult {
-            if !(3..=2048 * 2048).contains(&number) {
-                TestResult::discard()
-            } else {
-                let mut tree = initial();
-                for i in 1..(number-1) {
-                    tree.append_leaf(leaf(i+2)).expect("Failed to append");
-                }
+        #[test]
+        fn prop_parity(number in 3u32..=2048) {
+            let mut tree = initial();
+            for i in 1..(number-1) {
+                tree.append_leaf(leaf(i+2)).expect("Failed to append");
+            }
 
-                TestResult::from_bool(
-                    if number & (number - 1) == 0 {
-                        matches!(tree.root(), EntryLink::Stored(_))
-                    } else {
-                        matches!(tree.root(), EntryLink::Generated(_))
-                    }
-                )
+            if number & (number - 1) == 0 {
+                assert_matches!(tree.root(), EntryLink::Stored(_));
+            } else {
+                assert_matches!(tree.root(), EntryLink::Generated(_));
             }
         }
 
-        fn parity_with_truncate(add: u32, delete: u32) -> TestResult {
+        #[test]
+        fn prop_parity_with_truncate(
+            add_and_delete in (0u32..=2048).prop_flat_map(
+                |add| (Just(add), 0..=add)
+            )
+        ) {
+            let (add, delete) = add_and_delete;
```

Comment:
I believe this works:
```suggestion
            (add, delete) in (0u32..=2048).prop_flat_map(
                |add| (Just(add), 0..=add)
            )
        ) {
```

Reply:
I tried that first; the macro didn't like it.

Reply:
Hmm, maybe because of the `prop_flat_map`.